### DISCUSSION TITLE
Prerequisite notes

### DISCUSSION THREAD

**Discussion by IllustratedMan-code** [srid/emanote#358] [Ideas]
# Proposal
I propose some way to define prerequisite notes for a particular note. This way, if one wanted to understand a particular note, they could read the prerequisite notes. I feel that this is better than regular links because it provides a logical direction for a reader to follow when learning a topic.

For example, say I had a note explaining `derivatives`, I would list `algebra` as a prerequisite. I might also have a note explaining `physics` that lists `derivatives` as a prerequisite. This would provide a 1-way note relationship like so:
`algebra` -> `derivatives` -> `physics`

I'm not sure what kind of syntax would be most appropriate. Maybe a list under a specific headline:
``````markdown
# Prerequisites
- [[algebra]]
- [[trigonometry]]
``````

I feel this would be incredibly useful when paired with some type of node visualizer. A user could pick a topic (note) and be shown all of the prerequisite topics needed to learn it. I imagine this would look similar to the skill/technology tree concept commonly used in video games.

Technology tree in Civilization V
![image](https://user-images.githubusercontent.com/60514384/196488937-a246b391-88c0-4cef-8b4d-d702e9651726.png)

**Comment by srid** [comment]
@IllustratedMan-code This would be better done as a plugin or extension than a core feature. See https://github.com/EmaApps/emanote/discussions/250 & https://github.com/EmaApps/emanote/discussions/83

Normally you could use https://emanote.srid.ca/guide/export to get the relationships and create whatever visualizations you want, but in this case, you also want access to the Pandoc AST (to be able to tell which links are under a certain heading), which is currently possible through either of the following approaches:

- https://emanote.srid.ca/guide/lua-filters
- Extend Emanote in Haskell: https://github.com/EmaApps/emanote/issues/247

Perhaps write a problem statement and a concrete proposal of _how_ it can be best solved such that the more general case (extensibility) is also handled in some way?

**Reply by srid** [reply]
> I'm not sure what kind of syntax would be most appropriate.

Can we generalize on [folgezettel](https://emanote.srid.ca/guide/html-template/uptree) links? Emanote already supports `#[[Foo]]` or `[[Foo]]#` ... perhaps we can allow the user to overload that `#`?

**Reply by srid** [reply]
https://github.com/EmaApps/emanote/discussions/83

**Reply by IllustratedMan-code** [reply]
I'm happy to move this to #83 if needed.

I think a generalization on the folgezettel links is a great idea! In fact, it seems to be almost identical to what I was describing, the only difference is that uplinks seem to be one way (how do I go back down the tree?).
 I think a visualizer like I described could be fairly easy to implement as long as there is a good way to access the folgezettel link structure.

**Reply by srid** [reply]
> I think a visualizer like I described could be fairly easy to implement as long as there is a good way to access the folgezettel link structure.

Does the `export` command work? See screenshot. If we add support for extensible wiki-links (of which folgezettel becomes a special builtin kind), then this approach would also work for the extended links since the exported JSON would also be changed to include the extended wiki-links.

<img width="1118" alt="image" src="https://user-images.githubusercontent.com/3998/196978871-35a7733c-590a-403d-baa7-4f017a71cc29.png">

**Reply by IllustratedMan-code** [reply]
If the link type (folgezettel up and down) is contained in the exported json, then I don't see any problems with implementation.

**Reply by srid** [reply]
It should be, yes. I just need a concrete proposal for extending `?[[Foo]]?` such that both the `?` and their position (before or after) gets parsed appropriately. Do we treat *all* non-space letters immediately before/after the wiki-link to be special (of which `#` is just one possibility)? Though that would violate some sensible cases, like `Hyphenated-#[[Concepts]]`.

**Reply by IllustratedMan-code** [reply]
Just spitballing here, but you could use the first non-space character in most cases, or in the case of `{}` use the contents of that.

so all of these would be valid:

```
#[[Concepts]]
^[[Concepts]]  --  custom function assigned to ^
{myfunction}[[Concepts]] -- custom function
{#}[[Concepts]] -- Something like this could also be valid

[[Concepts]]#
[[Concepts]]^
[[Concepts]]{myfunction}
```

If the character right before *is* a space, then it could be removed from output so:

`Hyphenated-# [[Concepts]]` -> Hyphenated-#[Concepts](google.com) 

You could also take a page out of org-mode's book and have the link type in the bracket. I think this might make parsing less confusing.

```
[[#][Concepts]]
```
or maybe
```
[#[Concepts]]
```

Happy to write a proposal if you need.

**Reply by srid** [reply]
@IllustratedMan-code What other tools do you use to edit your Markdown notes? Obsidian? Vim? 

Something to keep in mind is compatibility with other tools and having your proposal veer towards it being more of an "extended Markdown standard" that other tools can adopt. This is how the wiki-link syntax itself was organically adopted. So something like `^[[Concepts]]` would be fitting, in that regard, than `[[^][Concepts]]` (because the latter will break wiki-link parsing of existing tools). 

You could also share your ideas in https://forum.obsidian.md/ and get some feedback on the syntax. Yes, an official proposal taking all this into consideration will be most welcome!

**Reply by IllustratedMan-code** [reply]
I jump back and forth between emacs and neovim typically, I can definitely see the need for cross compatibility.
I'll go to the obsidian forum for some feedback. Anywhere specific I should put the proposal?

**Reply by IllustratedMan-code** [reply]
Obsidian forum post: https://forum.obsidian.md/t/extended-link-syntax/46047

**Reply by IllustratedMan-code** [reply]
Unfortunately, no replies to the forum post. I'll go ahead and write a proposal here when I get the time.

**Comment by IllustratedMan-code** [comment]
#  High Level Proposal
This feature will require 2 sub-features to be implemented:
1. A tree visualizer like the technology tree pictured above
2. A way of injecting a hierarchical structure into the [export json](https://emanote.srid.ca/-/export.json)

## Tree Visualizer
I have some experience with javascript and the html canvas element. This would probably be the easiest method of implementation. We would simply fetch the export json and use it to generate the tree-like structure. 

## Export Json Injector
I'm not sure how hard this would be, but a custom function system would be very useful here. This would use the following syntax:
```
#[[Concepts]] -- custom function assigned to #
^[[Concepts]]  --  custom function assigned to ^
{myfunction}[[Concepts]] -- custom function
{#}[[Concepts]] -- Something like this could also be valid

[[Concepts]]#
[[Concepts]]^
[[Concepts]]{myfunction}
```
I'm not completely convinced by the curly bracket syntax, but I think it would be fairly easy to parse.


# Plan
I think it makes the most sense to work on the custom function system first, as it is immediately usable and applicable to other features/extensions. I'm very hesitant to extend the markdown syntax to prevent compatibility problems, but I don't think the proposed syntax is too intrusive.

@srid let me know what you think! I'm happy to go into more detail, but I will need to learn more about haskell and emanote's structure first.

**Reply by srid** [reply]
The `{..}` syntax can conflict with https://github.com/jgm/commonmark-hs/blob/master/commonmark-extensions/test/attributes.md (which is enabled in Emanote).

Is there a particular reason you need the custom function syntax (as opposed to a single-char syntax)?

**Reply by IllustratedMan-code** [reply]
I just thought it might be useful, I think single character is the right way to go if longer function names aren't needed.

**Reply by srid** [reply]
> I think single character is the right way to go if longer function names aren't needed.

Agreed. In general, let's try to keep things simple, and avoid complex things unless strictly necessary.

A good starting point into implementing this is to change this type into being more general (so it is not hardcoded to `#`):

https://github.com/srid/commonmark-wikilink/blob/72c880b5dbe8521f8c424c7bd39cb11b26e8dbb8/src/Commonmark/Extensions/WikiLink.hs#L174-L182

perhaps,

```haskell
data WikiLinkType
  = WikiLinkNormal
  | WikiLinkSuffixed Char
  | WikiLinkPrefixed Char
```

When it comes to using the forked library in emanote, you can look to https://github.com/srid/emanote/pull/399 for example/inspiration (see `flake.nix` changes).
